tests/test_warp.py

```diff
@@ -2146,3 +2146,113 @@ def test_issue2353bis(caplog):
+            np.arange(src.width), np.arange(src.height)
+        )
+
+        profile = src.profile
+        profile.update(
+            driver="GTiff",
+            height=800,
+            width=880,
+            transform=DST_TRANSFORM,
+            crs="EPSG:3857",
+        )
+
+    # Now alter the source's geotransform. If the geolocation arrays
+    # aren't used, we'll notice.
+    with rasterio.open(filename, "r+") as src:
+        src.transform = Affine.identity()
+
+    # Reproject the altered source file using provided geolocation
+    # arrays.
+    with rasterio.open(filename) as src:
+        with rasterio.open(tmp_path.joinpath("test.tif"), "w", **profile) as dst:
+            reproject(
+                rasterio.band(src, src.indexes),
+                rasterio.band(dst, dst.indexes),
+                src_crs=src.crs,
+                src_geoloc_array=np.stack((xs, ys)),
+                resampling=Resampling.bilinear,
+            )
+
+    with rasterio.open(tmp_path.joinpath("test.tif")) as dst:
+        out = dst.read(1)
+
+    # This value is specific to DST_TRANSFORM and an 800 x 880 file.
+    assert np.count_nonzero(out) in [464910]
+
+
+# Before GDAL 3.5.2 geoloc array files aren't recognized and this error
+# would be seen from the following tests:
+#
+# rasterio._err.CPLE_AppDefinedError: The transformation is already
+# "north up" or a transformation between pixel/line and georeferenced
+# coordinates cannot be computed for
+# MEM:::DATAPOINTER=137539856,PIXELS=791,LINES=718,BANDS=3,DATATYPE=Byte.
+# There is no affine transformation and no GCPs. Specify transformation
+# option SRC_METHOD=NO_GEOTRANSFORM to bypass this check.
+@pytest.mark.skipif(not gdal_version.at_least("3.6"), reason="Requires GDAL 3.6")
+def test_geoloc_warp_array(path_rgb_byte_tif, tmp_path):
+    """Warp an array using external geolocation arrays."""
+    with rasterio.open(path_rgb_byte_tif) as src:
+        xs, ys = src.transform * np.meshgrid(
+            np.arange(src.width), np.arange(src.height)
+        )
+        source = src.read()
+
+    output = np.zeros((3, 800, 880), dtype="uint8")
+
+    reproject(
+        source,
+        output,
+        src_crs=src.crs,
+        src_geoloc_array=np.stack((xs, ys)),
+        src_nodata=0,
+        dst_transform=DST_TRANSFORM,
+        dst_crs="EPSG:3857",
+        dst_height=800,
+        dst_width=880,
+        resampling=Resampling.bilinear,
+    )
+
+    # This value is specific to DST_TRANSFORM and an 800 x 880 array.
+    assert np.count_nonzero(output[0]) in [464910]
+
+
+@pytest.mark.skipif(not gdal_version.at_least("3.6"), reason="Requires GDAL 3.6")
+def test_geoloc_warp_array_subsampled(path_rgb_byte_tif, tmp_path):
+    """Warp an array using subsampled external geolocation arrays."""
+    with rasterio.open(path_rgb_byte_tif) as src:
+        xs, ys = src.transform * np.meshgrid(
+            np.arange(0, src.width, 10), np.arange(0, src.height, 10)
```

Comment:
I'm kind of surprised this works. I expected the transformer to need `PIXEL_STEP` and `LINE_STEP` to function properly. Any idea whats going on? Is GDAL assigning a default value?

Reply:
@pl-kevinwurster PIXEL_STEP and LINE_STEP defaults are documented in https://gdal.org/api/gdal_alg.html#_CPPv432GDALCreateGenImgProjTransformer212GDALDatasetH12GDALDatasetHPPc.